test/1.2/statements/T_if_ambiguity.dml

```diff
@@ -0,0 +1,60 @@
+/*
+  © 2023 Intel Corporation
+  SPDX-License-Identifier: MPL-2.0
+*/
+dml 1.2;
+
+device test;
+
+// This tests both that GCC doesn't emit Wparantheses and the behavior of
+// the DML parser.
+// Although GCC emitting Wparantheses could be considered desirable for the
+// patterns this test exemplifies, even if braces were inserted to disambiguate
+// the generated C would be identical due to `mkCompound` collapse.
+// Thus Wparantheses must be avoided in general.
```

Comment:
if this mainly covers that the test looks good to GCC, then `/// COMPILE-ONLY` should suffice?

Also, looks a bit like a smoke test; did you consider testing on a ctree_test level? E.g., `toc()` difficult corners into a StrOutput and compare the result character-by-character. The upside would be a test case where you can see the brace generation difficulties visually; the downside would be that the test would go more technical and brittle.

Reply:
The runtime component is to test the behavior of the DML parser -- to which `if` each `else` is associated with. If we're only interested testing for `Wparantheses` then `/// COMPILE-ONLY` would suffice, yes.

This is not really a smoke test: what we care about, explicitly, is that GCC doesn't emit Wparantheses for uses of if-else, and that can only occur in very specific circumstances. See https://github.com/gcc-mirror/gcc/blob/master/gcc/testsuite/gcc.dg/Wparentheses-12.c and https://github.com/gcc-mirror/gcc/blob/master/gcc/testsuite/gcc.dg/Wparentheses-13.c . In our case, this can only possibly occur with DML statements for which:
1. The definition of `toc_stmt()` itself uses `toc_stmt()` of some arbitrary parameter statement
2. The definition of `toc_stmt()` doesn't enclose the inner use of `toc_stmt()` in braces

The relevant statements for this is applicable are rather easy to find. The only one which satisfies both and isn't covered here is `DoWhile`, and that's because it eliminates ambiguity by virtue of the `while (...);`.

Reply:
In some sense, leveraging `ctree_test` would be _more_ of a smoke test because then we wouldn't run it past GCC.

Reply:
@mandolaerik Reminder. Are you convinced by my arguments?

Reply:
Sorry I forgot this. I just wanted to say there's something about the test that itches a bit, and I gave my best guess why it does. But itchy tests with good coverage are OK; we can improve the test later if it ever becomes a real problem.